### ISSUE TITLE
dealing with circular dependency in sst v3

### ISSUE THREAD

**Issue by leo-the-nardo** [anomalyco/sst#5155]
there is any way to deal with circular dependencies in sst v3?
i have recorrent scenarios where  i need to send properties of a resource to another dependent resource. a good example is that my front end needs some variables of my back-end ,and vice-versa, also a scenario where i need the Router id to invalidate the cache from the application itself.
i tried some of the [solutions provided by pulumi](https://www.pulumi.com/blog/exploring-circular-dependencies/) but i was unable to apply them in sst v3

**Comment by goto3** [comment]
you sholuld analize why you need the backend to deppend on frontend variables, in any case, a solution could be either creating a separate file with that variables so that both BE and FE depend on it. Another solution should be the use of SSM setting variables through Pulumi

**Comment by cdelmoral** [comment]
Interested in this. Predefining the variable so that both resources can depend on it make sense, but what happens when they are generated by AWS?

Example:
- Create a cognito user pool client
  - It generates a client id/secret
  - It requires a callback url
- Create an application (e.g. a backend) that uses the cognito user pool client
  - It generates a cloudfront distribution url
  - It requires the user pool client id/secret to inject it as an environment variable

**Comment by zammitjames** [comment]
I'm confused - why can't the callback url be defined in the infra itself based on the frontend dns config?

**Comment by cdelmoral** [comment]
> I'm confused - why can't the callback url be defined in the infra itself based on the frontend dns config?

When using a custom domain that makes sense.

But I'm thinking about dev/ci environments where we would use the auto-generated cloudfront url instead. Then we have this circular dependency where:
1. The frontend needs the cognito user pool client in order to link it
2. The cognito user pool client needs the frontend callback url

**Comment by leo-the-nardo** [comment]
> you sholuld analize why you need the backend to deppend on frontend variables, in any case, a solution could be either creating a separate file with that variables so that both BE and FE depend on it. Another solution should be the use of SSM setting variables through Pulumi

in theory, SSM parameter stores partially solves the problem ,but this comes with complexity for the application side, and also add overhead and latency in lambda or serverless scenarios, sometimes +100ms to execution - which at the end of the day turns into a cost

**Comment by mschinis** [comment]
I am also having issues with the above issue, as I am currently trying to create two lambda functions like the following, where the API and Auth need to know each other's endpoint:

- Auth -> Api: Because we need to retrieve user information from the API during the authentication process
- Api -> Auth: So we can initialise the openauth client, in order to check the validity of a token.

```ts
        const api = new sst.aws.Function("Api", {
            url: true,
            handler: "apps/api/src/handlers/aws-lambda.handler",
            environment: {
                DATABASE_URI: $interpolate`...`,
            }
        }, {
            dependsOn: [db]
        });

        const auth = new sst.aws.Auth("Auth", {
            authorizer: {
                handler: "apps/auth/src/handlers/aws-lambda.handler",
                environment: {
                    API_URI: $interpolate`${api.url}`,
                    API_TOKEN: 'something'
                },
            },
        });
```

Looking at the pulumi documentation and through the sst codebase, it looks like the [Option 4](https://www.pulumi.com/blog/exploring-circular-dependencies/#option-4-allow-a-resource-to-update-its-goal-state) should be easy to implement in the `sst.config.ts` file if the underlying lambda function is exposed?

**Comment by jayair** [comment]
@mschinis Which underlying Lambda? The auth or the api?

**Comment by john-griffin** [comment]
Yeah we also hit this. We have a backend API that also sends emails with links to frontend app. So the API needs the frontend url and the frontend needs the API url.

**Comment by claridgicus** [comment]
I have this issue also.

**Comment by jayair** [comment]
This doesn't address everybody's use case but we recently added this for functions, it allows you to lazily add an env var for a function to sidestep some circular dependency issues. I'll need to create a separate doc for this.

https://sst.dev/docs/component/aws/function/#addenvironment

**Comment by john-griffin** [comment]
Ah yeah thats great for functions! Would this be possible to implement on Service components or is that trickier due to baking in the env vars?

**Comment by jayair** [comment]
I'll need to check.

**Comment by wilfredjonathanjames** [comment]
I have a Function with success and failure Functions. In some cases these Functions can invoke the first Function again. This requires bidirectional linking between the first and the other two. Besides forcing the permissions with Linkable somehow, it's not clear how to proceed.

**Comment by wilfredjonathanjames** [comment]
I ended up setting permissions with pulumi and injecting the function name in environment rather than `Resource`. Extremely verbose but works fine

```typescript
const Func__Failure = new sst.aws.Function("Func__Failure", {
  handler: "src/lambdas/func.onFailure",
  retries: 0,
})
const Func = new sst.aws.Function("Func", {
  handler: "src/lambdas/func.handler",
  retries: 0,
  link: [Func__Failure],
})

// inject Func name into Func__Failure for invocation
Func.name.apply((name) => {
  Func__Failure.addEnvironment({
    FUNC__NAME: name,
  })
})

// call Func__Failure on Func failure
new aws.lambda.FunctionEventInvokeConfig("FuncFunctionConfig", {
  functionName: Func.name,
  maximumRetryAttempts: 0,
  destinationConfig: {
    onFailure: {
      destination: Func__Failure.arn,
    },
  },
})

// create a lambda invoke policy for Func
const Func__InvokePolicy = new aws.iam.Policy("Func__InvokePolicy", {
  name: "Func__InvokePolicy",
  policy: Func.arn.apply(async (arn) => {
    return aws.iam
      .getPolicyDocument({
        statements: [
          {
            actions: ["lambda:InvokeFunction"],
            effect: "Allow",
            resources: [arn],
          },
        ],
      })
      .then((example) => example.json)
  }),
})

// bind the invoke policy to the Func__Failure execution role
new aws.iam.RolePolicyAttachment(
  "Func__InvokePolicy__RoleAttachment__FailureFunction",
  {
    role: Func__Failure.name.apply(async (functionName) => {
      return (
        await aws.lambda.getFunction({
          functionName,
        })
      ).role.match(/\/(.*)$/)![1] // pull role name off the end of the ARN
    }),
    policyArn: Func__InvokePolicy.arn,
  },
)
```
